### PR TITLE
test: extract manual dissemination performance tooling

### PR DESCRIPTION
**Pull request by ReubenBond**
## Problem

The dissemination feature in #10236 needs focused correctness coverage, while performance experiments need independently selected runtime revisions and manually chosen resource budgets.

## Solution

Extract the measurement harness from `3fc0a4610cd2eb5feef27e958fcca59658aaae47` into `test/Dissemination.PerformanceHarness`, with a `workflow_dispatch`-only workflow and bounded PowerShell entrypoint. The workflow selects an explicit candidate repository/ref and builds it alongside the pinned original `6739589254b746a8790cf53524e6abe372bb53d4` in isolated checkouts. Only Silo/Shared support code is copied from the tooling checkout; each worker resolves runtime references within its selected checkout.

The controller measures original, candidate-off and candidate-on paths across stable/churn/partition workloads, reverses paired repetition order, checks exact state equality and compares equal binary/environment/process count/offered work. Reports preserve request versus logical-value counts, CPU/allocation/socket costs, convergence distributions and per-node evidence. Partition recovery retains the #11243 endpoint-drain and bidirectional-ack barrier, including readiness costs before one publication round. Documentation covers revision selection, artifact paths, resource limits and local `-SkipBuild` replay.

The default workload remains closed-loop. Optional synchronized and evenly staggered stable workloads offer 1 Hz per producer on worker-local monotonic schedules, bounded to 3..32 processes and 3..30 seconds. Their cadence is independent of controller polling. Reports preserve planned/actual publication times, missed periods and overruns, observed/unobserved value pairs, polling latency upper bounds, and exact final latest-state convergence. Candidate aggregation fanout uses the optional public aggregation-specific setting with its effective routing clamp; older candidates retain an explicitly labeled legacy capability path.

## Scope and rationale

This is standalone tooling against main. The diff adds the performance harness, scripts and manual workflow, plus links for private-provider regressions in `Orleans.Runtime.Tests` and lightweight Linux/Windows CI checks. Production runtime, production package/framework configuration and the feature's correctness harness remain unchanged. Performance workloads run only through explicit manual invocation. Shared-runner loopback results describe the selected control-plane workload; same-candidate off/on pairs isolate enablement.

The runner constructs platform-native paths and validates existing destination parents before copying workers. File membership cleanup follows the runtime's non-active-status and `EffectiveUpdateTime` semantics. Ordinary CI covers stubbed replay paths and persisted-provider cleanup; focused controller cases use the existing centrally versioned fake-time testing package for deterministic 1 Hz scheduling checks.

The initial bounded 4/8-process dispatch [35118950976](https://github.com/ReubenBond/orleans/actions/runs/35118950976) remains queued with zero jobs at bootstrap head `a4bd25c4030daae5cbf7cecbca38a24c5050eab4`; fork Actions are disabled. The temporary workflow was restored before opening this PR. Current tooling head `fb0762e0787f60e7958877305cf80ed17316ab67` has green [ordinary CI](https://github.com/dotnet/orleans/actions/runs/35206583735), while exact-head real-worker partition and fixed-rate evidence remains pending.

The shared #11243 recovery mechanism has real-process evidence in #10236 at `48302ff632bbc7f99eadc0ce366b08c2ca4d9b88`: [compatibility run 35180410047](https://github.com/dotnet/orleans/actions/runs/35180410047) passed 10 process cases and 8 harness cases, with 52 process lifetimes, zero forced kills, and 18 healed publication rounds across 4/8-silo original-off, candidate-off, and candidate-on scenarios. These counts were verified from artifact `dissemination-compatibility-results` (`10480451697`); they establish parent-harness coverage, with this extracted tooling's real-worker validation still pending.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/orleans/pull/11272)

### CODE REVIEW

File: test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        state.Rows.Clear();
+        state.Version++;
+        return true;
+    }, write: true, cancellationToken);
+
+    public Task CleanupDefunctSiloEntriesAsync(DateTimeOffset beforeDate, CancellationToken cancellationToken) => Access(state =>
+    {
+        foreach (var key in state.Rows.Where(pair =>
+            pair.Value.Entry.Status == (int)SiloStatus.Dead
+            && pair.Value.Entry.IAmAliveTime < beforeDate.UtcDateTime).Select(pair => pair.Key).ToArray())
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs]: Cleanup now checks `Status != SiloStatus.Active` and uses the selected runtime's `MembershipEntry.EffectiveUpdateTime`. Added persisted-provider regressions for every non-active status and startup, heartbeat, and suspect-vote timestamps immediately before, exactly at, and after the cutoff. The ten cases passed on Windows/.NET 10 and Linux/.NET 8. They are linked into the ordinary BVT suite; performance workloads remain manually invoked.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    [ValidateRange(0, 9)]
+    [int] $GCConserveMemory = 0,
+    [switch] $SkipBuild
+)
+
+$ErrorActionPreference = 'Stop'
+Set-StrictMode -Version Latest
+$baselineSha = '6739589254b746a8790cf53524e6abe372bb53d4'
+$root = [System.IO.Path]::GetFullPath((Join-Path $PSScriptRoot '..\..'))
+$harnessSource = Join-Path $root 'test\Dissemination.PerformanceHarness'
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1]: Changed the multi-component paths to separate `Join-Path` segments, including the default artifact directory. The validated artifact name is normalized so either slash spelling resolves to the same directory. Added successful `-SkipBuild` path fixtures which stub worker execution and check runner selection and persisted result paths. All 17 boundary cases and both replay-path cases pass on Windows and Linux CI.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+    public Task DeleteMembershipTableEntriesAsync(string clusterId, CancellationToken cancellationToken) => Access(state =>
+    {
+        if (!StringComparer.Ordinal.Equals(clusterId, configuration.ClusterId))
+        {
+            throw new InvalidOperationException("The harness provider cannot delete another cluster.");
+        }
+
+        state.Rows.Clear();
+        state.Version++;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs:71]: Fixed the delete mutation in fcd7ba2654: clearing the rows now advances the table ETag as well as its version. `DeleteInvalidatesThePreviousTableEtag` reopens the persisted table and verifies stale inserts/updates are rejected while a write using the new version succeeds.

The heartbeat site follows a different contract: `IMembershipTable.UpdateIAmAlive` explicitly permits retaining the row ETag and requires leaving the table version unchanged. I preserved that behavior and added `HeartbeatPreservesTableVersionAndOtherColumns` to verify that only the liveness timestamp changes. All 12 provider regression cases passed in Linux/.NET 8 BVT CI at this commit.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    $current = $Path
+    while ($current -and $current -ne $root) {
+        if (Test-Path -LiteralPath $current) {
+            $item = Get-Item -LiteralPath $current -Force
+            if ($item.Attributes -band [System.IO.FileAttributes]::ReparsePoint) {
+                throw "Artifact and runtime paths must use ordinary directories: $current"
+            }
+        }
+        $current = Split-Path -Parent $current
+    }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1]: Added an explicit normalized repository-containment check before traversing parents, plus an empty/unchanged-parent termination guard. Outside paths now fail with the repository-boundary diagnostic. Regression cases cover the repository itself, a not-yet-created descendant, its parent, the filesystem root, and a sibling with the same prefix.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Silo/Program.cs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        {
+            using var startup = CancellationTokenSource.CreateLinkedTokenSource(lifetime.Token);
+            startup.CancelAfter(TimeSpan.FromSeconds(90));
+            await host.StartAsync(startup.Token);
+            var publisher = host.Services.GetRequiredService<DeploymentLoadPublisher>();
+            // Commands own publication cadence; lifecycle shutdown retains its disposable timer reference.
+            await publisher.RunOrQueueTask(() =>
+            {
+                var field = typeof(DeploymentLoadPublisher).GetField("_publishTimer", BindingFlags.Instance | BindingFlags.NonPublic)!;
+                ((IDisposable)field.GetValue(publisher)!).Dispose();
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/Program.cs]: The publisher-boundary helper now validates and caches `_publishTimer` as an IDisposable-compatible field, with the same targeted runtime-adapter diagnostic for missing/incompatible fields. The post-startup timer lifecycle guarantee remains intact: pausing disposes the timer and retains its field reference for shutdown. Added regressions for missing/wrong-type fields and a compatible publication which verifies the original cancellation token and retained disposed-timer reference.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Tests/ProcessCluster.cs [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+    string Runtime,
+    string SourceRevision,
+    string AssemblyName,
+    string AssemblyVersion,
+    string Sha256,
+    Dictionary<string, AssemblyManifest> Assemblies);
+
+internal sealed class ProcessCluster : IAsyncDisposable
+{
+    public const string Baseline = "6739589254b746a8790cf53524e6abe372bb53d4";
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Tests/ProcessCluster.cs:21]: Kept the immutable original-runtime pin as an independent check in the controller rather than trusting a downloaded selection manifest to redefine the baseline. Added lightweight consistency checks which require the script, controller, workflow checkout, methodology and README to agree on that pin. This catches drift while preserving the exact-baseline validation boundary.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    if ($LASTEXITCODE -ne 0 -or $changes) {
+        throw "Selected runtime checkout must be pristine before copying worker support: $Directory"
+    }
+}
+
+function Copy-Worker([string] $Checkout) {
+    $destination = Join-Path $Checkout 'test' 'Dissemination.PerformanceHarness'
+    if (Test-Path -LiteralPath $destination) {
+        throw "Refusing to overwrite an existing runtime harness: $destination"
+    }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1:111]: Checked the immutable `3fc0a4610cd2eb5feef27e958fcca59658aaae47` tree: it contains `test/Dissemination.IntegrationHarness`, not `test/Dissemination.PerformanceHarness`. The copy destination therefore does not collide for the documented candidate. Added an executable copy fixture with an existing IntegrationHarness directory; it verifies that only the tooling's Silo/Shared sources are copied byte-for-byte, the existing directory is preserved, and a real PerformanceHarness collision is still rejected. Keeping the overwrite guard prevents silently mixing selected-runtime support with this tooling revision.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+    Task<MembershipTableData> IMembershipTable.ReadAll() => ReadAllAsync(CancellationToken.None);
+
+    Task<bool> IMembershipTable.InsertRow(MembershipEntry entry, TableVersion tableVersion) =>
+        InsertRowAsync(entry, tableVersion, CancellationToken.None);
+
+    Task<bool> IMembershipTable.UpdateRow(MembershipEntry entry, string etag, TableVersion tableVersion) =>
+        UpdateRowAsync(entry, etag, tableVersion, CancellationToken.None);
+
+    Task IMembershipTable.UpdateIAmAlive(MembershipEntry entry) => UpdateIAmAliveAsync(entry, CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs:32]: The `#if NEW_RUNTIME` branch already contains all eight legacy forwarding members as explicit `IMembershipTable` implementations (lines 12–33), including `InitializeMembershipTable(bool)` and `UpdateRow(...)`. They forward to the cancellation-aware methods with `CancellationToken.None`. Explicit interface implementations satisfy the abstract interface contract just as the public forwarding methods in `#else` do; adding another set would duplicate those implementations. Verified this against both the current file and the selected candidate's interface; no implementation change is needed.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Silo/Program.cs [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+using Orleans.Runtime.Scheduler;
+using Orleans.TestingHost;
+
+namespace Orleans.Dissemination.PerformanceHarness;
+
+internal static class Program
+{
+    private static readonly MethodInfo PublishMethod = typeof(DeploymentLoadPublisher).GetMethod(
+        "PublishStatistics", BindingFlags.Instance | BindingFlags.NonPublic)!;
+
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/Program.cs]: Added a shared publisher-boundary helper which requires the exact `Task PublishStatistics(CancellationToken)` signature. Missing/incompatible methods now produce an `InvalidOperationException` naming the selected assembly, type, required member and adapter update action. Focused cases cover missing methods, incorrect parameters and an incompatible return type.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Orleans.Runtime.Tests/Orleans.Runtime.Tests.csproj [hunk 10]
```diff
@@ -9,6 +9,12 @@
     <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Include="..\Dissemination.PerformanceHarness\Silo\FileMembershipTable.cs" Link="DisseminationPerformance\FileMembershipTable.cs" />
+    <Compile Include="..\Dissemination.PerformanceHarness\Shared\Protocol.cs" Link="DisseminationPerformance\Protocol.cs" />
+    <Compile Include="..\Dissemination.PerformanceHarness\ProviderTests\FileMembershipTableTests.cs" Link="DisseminationPerformance\FileMembershipTableTests.cs" />
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Runtime.Tests/Orleans.Runtime.Tests.csproj:16]: Keeping the three explicit source links. The provider must compile inside the existing `Orleans.Runtime.Tests` friend assembly for private runtime access, while each measured worker compiles the same support against its independently selected runtime checkout. A separately referenced shared assembly would change that boundary and require additional friend access or runtime dependencies. Explicit file links also make a moved/missing source a build failure rather than silently excluding it through a glob.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Tests/ScalingComparison.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            .GroupBy(sample => (sample.Size, sample.Scenario, sample.Repetition)))
+        {
+            if (group.Count() == 1)
+            {
+                incomplete++;
+                continue;
+            }
+
+            var disabled = group.Single(sample => sample.RuntimePath == "CurrentDisabled");
+            var enabled = group.Single(sample => sample.RuntimePath == "CurrentEnabledSupported");
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Tests/ScalingComparison.cs:26]: Keeping this as written. LINQ `GroupBy` has already materialized each group, and its grouping exposes a count that `Count()` can obtain directly. A valid group contains just the two off/on samples, so these `Single` calls scan a tiny in-memory collection and also reject duplicated or missing paths. `ToArray()` would introduce another allocation without avoiding repeated source enumeration.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/workflows/dissemination-performance.yml [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+          ref: 6739589254b746a8790cf53524e6abe372bb53d4
+          path: .dissemination-runtime-original
+          persist-credentials: false
+      - name: Selected candidate runtime checkout
+        uses: actions/checkout@93cb6efe18208431cddfb8368fd83d5badbf9bfd # v5
+        with:
+          repository: ${{ inputs.candidate_repository }}
+          ref: ${{ inputs.candidate_ref }}
+          path: .dissemination-runtime-candidate
+          persist-credentials: false
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/workflows/dissemination-performance.yml:85]: Retaining explicit repository/ref selection: executing a chosen runtime revision is the purpose of this manually dispatched tool, and GitHub gates dispatch on repository write access. An owner-name allowlist would still allow arbitrary branch contents within those repositories while excluding other deliberately selected forks. The workflow uses an ephemeral hosted runner, read-only contents permission, disabled checkout credentials and no configured secrets. Expanded the README to state that the dispatcher authorizes the selected code and its build dependencies, including runner network access, and should inspect them before dispatch. The requested ref and resolved SHA remain recorded for audit.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    foreach ($part in @('Silo', 'Shared')) {
+        $partDestination = Join-Path $destination $part
+        New-Item -ItemType Directory -Path $partDestination -Force | Out-Null
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1:114]: Fixed in 33fffa2965. `Copy-Worker` now validates its full destination path and every existing parent before creating directories or copying files, so linked `test` directories inside an otherwise valid checkout are rejected.

Added four regressions covering links at both `test` and `test/Dissemination.PerformanceHarness`, with existing and dangling targets. They verify the expected boundary diagnostic and that targets are neither modified nor created. The first case reproduced the escape before the fix; all script checks now pass locally and in the Linux/Windows CI steps.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        throw "Path is outside the repository boundary: $Path"
+    }
+    $comparison = if ($IsWindows) { [StringComparison]::OrdinalIgnoreCase } else { [StringComparison]::Ordinal }
+    while (![string]::Equals($current, $root, $comparison)) {
+        if (Test-Path -LiteralPath $current) {
+            $item = Get-Item -LiteralPath $current -Force
+            if ($item.Attributes -band [System.IO.FileAttributes]::ReparsePoint) {
+                throw "Artifact and runtime paths must use ordinary directories: $current"
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1]: Changed the traversal in 80df772632 to inspect each entry directly with `Get-Item -Force -ErrorAction Stop`. Only `ItemNotFoundException` is treated as an absent path; permission and other inspection failures propagate. The existing four linked-parent cases cover both existing and dangling targets, and a new regression verifies access failures are propagated at the requested path. All script checks pass locally.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: .github/scripts/dissemination-performance.ps1 [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    GCConserveMemory = $GCConserveMemory
+    StartUtc = [DateTimeOffset]::UtcNow
+    ResultsDirectory = $results
+} | ConvertTo-Json -Depth 5 | Set-Content -LiteralPath (Join-Path $results 'invocation.json')
+
+Push-Location $root
+try {
+    Invoke-DotNet @(
+        $runner, '--filter-class', 'Orleans.Dissemination.PerformanceHarness.MeasurementTests',
+        '--minimum-expected-tests', '50', '--report-trx',
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply .github/scripts/dissemination-performance.ps1:297]: Kept the gate at 50. The actual MTP run with `--filter-class '*MeasurementTests' --minimum-expected-tests 50 --report-trx` passes all 50 cases. The TRX breakdown is 15 measurement cases + **24** open-loop cases + 6 publisher API cases + 5 aggregation cases. The open-loop partial includes the `OpenLoopArmingIsSingleUse` fact in addition to its other expanded cases, so reducing the minimum would weaken the existing check.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        int FaultZone,
+        DateTime StartTime,
+        DateTime IAmAliveTime,
+        Dictionary<string, DateTime> Suspects)
+    {
+        public static EntryData From(MembershipEntry entry) => new(
+            entry.SiloAddress.ToParsableString(), (int)entry.Status, entry.ProxyPort,
+            entry.HostName, entry.SiloName, entry.RoleName, entry.UpdateZone, entry.FaultZone,
+            entry.StartTime, entry.IAmAliveTime,
+            entry.SuspectTimes?.ToDictionary(pair => pair.Item1.ToParsableString(), pair => pair.Item2) ?? []);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Dissemination.PerformanceHarness/Silo/FileMembershipTable.cs]: Fixed in fb0762e078. Suspect votes are persisted as an ordered nullable list of address/time records, preserving repeated voters, timestamps, and null-versus-empty state. Added persisted insert/update round-trip tests with duplicate votes and verified cleanup uses the newest vote with an exclusive cutoff. All 15 provider regression cases pass on Linux and Windows under both .NET 8 and .NET 10; current-head CI is green.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

**Comment by ReubenBond** [comment]
Moving the #11243 investigation onto this PR and closing the separate issue because the failure and correction are confined to the open dissemination tooling/feature stack.

The captured legacy-lane timeout was a harness synchronization failure at partition healing: a one-way `UpdateRuntimeStatistics` publication was dropped when the old connection terminated, and the paused publisher provided no later publication opportunity during the exact-state assertion. This PR retains the endpoint-drain plus acknowledged bidirectional-control barrier before the single publication round. The parent-harness run passed 10 process and 8 harness cases across 52 process lifetimes and 18 heals with zero forced kills.

Remaining acceptance belongs here: exact-tooling-head real-worker validation is still pending because fork Actions are disabled. The unobserved scheduling hypothesis involving a connection queued before middleware admission also remains PR review scope; it has no observed reproducer.
